feat(avm)!: Bytecode Hashing

MirandaWood opened the pull request:
## Constraining Bytecode Hashing

### Overview

Following elsewhere in the repo, we constrain that the bytecode commitment is the poseidon2 hash of the bytecode (encoded as fields) prepended by the special separator (`GENERATOR_INDEX__PUBLIC_BYTECODE`):
```
hash = poseidon2([GENERATOR_INDEX__PUBLIC_BYTECODE, FF(bytecode[0...31]), FF(bytecode[31...62]), ... ])
```

See `encode_bytecode` and `compute_public_bytecode_commitment` for C++ logic (unchanged by this PR).

### PIL

The general idea is that each row of `bc_hashing` corresponds to a row of `poseidon2`. It is the 'link' between `bc_decomposition` and `poseidon2`, proving that the hash given by `output_hash` is indeed a valid hash of the exact bytecode from `bc_decomposition` (with no extra or missing info), prepended by `GENERATOR_INDEX__PUBLIC_BYTECODE`.

The decomp circuit already helpfully sets the packed field every 31 bytes (looking forward), where `sel_packed = 1` and `packed_field` is the field given by packing the bytes from `pc` to `pc + 31` (starting at 0). Since each poseidon permutation takes in a chunk of 3 fields, each row of `bc_hashing` handles 3 packed fields. For now, I'm looking up each separately (for `i = 0, 1, 2`):
```
    #[GET_PACKED_FIELD_i]
    sel { pc_index_i, bytecode_id, packed_fields_i }
    in
    bc_decomposition.sel_packed { bc_decomposition.pc, bc_decomposition.id, bc_decomposition.packed_field };
```
where (generally) each `pc_index_i` is simply `pc_index_{i - 1} + 31`. Then each field is looked up into the hasher:

```
    #[POSEIDON2_HASH]
    sel { latch, packed_fields_0, packed_fields_1, packed_fields_2, output_hash }
    in poseidon2_hash.sel { poseidon2_hash.end, poseidon2_hash.input_0, poseidon2_hash.input_1, poseidon2_hash.input_2, poseidon2_hash.output };
```

Easy! Oh wait. Edge cases.

#### Bytecode Start

At the beginning of a new bytecode, we have `START = 1` and `pc_index = 0`. We must constrain that the first field hashed is the prepended separator:

```
    #[START_IS_SEPARATOR]
    START * (packed_fields_0 - constants.GENERATOR_INDEX__PUBLIC_BYTECODE) = 0;
```
 However, this means that `packed_fields_0` does not actually exist as a packed field in `bc_decomposition`, so the lookup `GET_PACKED_FIELD_0` will fail. To rectify this, at `START = 1`:
 - We have `sel_not_start` = 0 and skip the `GET_PACKED_FIELD_0` lookup
 - We set `pc_index_1 = 0`, so `GET_PACKED_FIELD_1` reads the first encoded field of the bytecode from `bc_decomposition` (and => `pc_index_2 = 31` and reads the second)
 - Instead of constraining that `pc_index' = pc_index + 93` (usual row), we only increment by 62
 - The poseidon lookup remains the same

#### Bytecode End

As before, once we read the end of some bytecode, we set `latch = 1` and constrain that the next row is a `start` (if exists). We have two main things to consider here:
 - Constrain that when `latch = 1`, we are at the end of decomposed bytes given by `bc_decomposition` (and read all of them)
 - Deal with padding in `poseidon2`
Both are dealt with in the `PADDING` section, since both require asserting we have 'trailing zeros' and have correctly read/set the packed fields.

Now that each row handles 3 fields, we may have 0, 1, or 2 padding fields which will not exist in `bc_decomposition`. To avoid the lookups bricking, I added two selectors:
 - `sel_not_padding_1` <==> `packed_field_1` is not a padding value => switch on the `GET_PACKED_FIELD_1` lookup
 - `sel_not_padding_2` <==> `packed_field_2` is not a padding value => switch on the `GET_PACKED_FIELD_2` lookup

(Note that `packed_field_0` can never be a padding value and must always be processed, and all padding selectors are gated by `latch`)

This now adds an attack vector where we could incorrectly claim real values are padding values in order to skip the lookup and => skip hashing bytecode fields. To avoid this I added `pc_at_final_field` which is the value of `pc` at the beginning of the last field e.g. if `packed_fields_1` is the last bytecode field, then `pc_at_final_field = pc_index_1`. This is constrained by the relation `PADDING_CORRECTNESS`.

From here, the simplest way to prove everything we need for padding/end of bytecode is to check that the `bc_decomposition.bytes_remaining` at `pc_at_final_field` is `< 31`. But this needs an evil range check. I don't want an evil range check on top of the 3 packed field lookups. Hence the following madness.

The decomposition trace processes windows of (currently) 37 bytes, and forces these bytes to be zero if they overshoot the end of the bytecode. Iff we have less than 37 bytes remaining to process, `bc_decomposition.sel_windows_gt_remaining = 1` => we can easily check that `bytes_remaining < 37` without a range check by just looking up this selector.

What about the remaining 6 bytes, you ask? These are all stored in `bc_decomposition.bytes_pc_plus_i`, so the final lookup becomes this monstrosity:
```
   #[CHECK_FINAL_BYTES_REMAINING]
    latch {
        pc_at_final_field,
        bytecode_id,
        sel /* =1 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */
    } in bc_decomposition.sel_packed {
        bc_decomposition.pc,
        bc_decomposition.id,
        bc_decomposition.sel_windows_gt_remaining,
        bc_decomposition.bytes_pc_plus_31,
        bc_decomposition.bytes_pc_plus_32,
        bc_decomposition.bytes_pc_plus_33,
        bc_decomposition.bytes_pc_plus_34,
        bc_decomposition.bytes_pc_plus_35,
        bc_decomposition.bytes_pc_plus_36
    };
```

But hey, we know we're at the final field at `pc_at_final_field` and there are no bytes remaining without a range check! Note that if the bytecode ends in valid zeros, I don't think we need to differentiate between these and 'padding' non-existent zeros because the final poseidon hash result would be the same.

To cover the poseidon considerations, we just need to make sure that the padding fields are indeed zero (`PADDED_BY_ZERO_i`) and if we claim that `packed_field_1` is a padding value, then so is `packed_field_2` (i.e. we can't have `[last_field, padding, some_crap]` at the end of a bytecode, `PADDING_CONSISTENCY`). If we have any padding at all, then `PADDING_2 = 1 - sel_not_padding_2` must be on, so I use this as the general padding selector when required.


#### Bytecode Start & End

Oh yes. These things can both be true. Isn't that fun! You might notice some ugly complexities in relations and this is probably why.

A much nicer way of showing `bytes_remaining < 31` without a range check would be to set some `pc_index_to_lookup = pc_at_final_field - 6`, then show that at `pc_index_to_lookup` that `bc_decomposition.sel_windows_gt_remaining = 1`. However, if both `latch` and `start` are true here then we have `pc_index = 0` and this would underflow.

I also made sure to use the `pc_index_i` columns to work out `pc_at_final_field` because they take into consideration the shift we do at `start` for the separator. Otherwise we could use something like `31*padding_amount` which would be a lot prettier (alas).

### TODOs/Qs

 - Should we explore using shifts to remove the 3 lookups into `bc_decomposition` for the packed fields?
 - Since the padding/latch conditions (I think?) constrain that we hash all bytes, the `bytecode_length` event property is unused, remove?
 - Is it worth converting the padding columns to a numerical `padding_amount` (=0, 1, or 2) and looking this up into `poseidon_2`?

MirandaWood commented on barretenberg/cpp/src/barretenberg/vm2/common/instruction_spec.hpp:
```diff
@@ -12,6 +12,12 @@
 namespace bb::avm2 {
 
 constexpr size_t NUM_OP_DC_SELECTORS = 18;
+constexpr uint32_t DECOMPOSE_WINDOW_SIZE = 37; // Max size of instruction in bytes.
+// We do not use a static constant to compute this value from WIRE_INSTRUCTION_SPEC
+// because we do not want it to be changed transparently. Namely, if this constant
+// is changed, we need to make significant changes in pil/vm2/bytecode/bc_decomposition.pil
+// and pil/vm2/bytecode/instr_fetching.pil.
+// We rely on the unit test InstructionSpecTest.CheckDecomposeWindowSize to detect if this constant is changed.
```
This (and the test.cpp and decomp test) change just brings in changes from #16520 relevant to window size.

MirandaWood commented on noir-projects/aztec-nr/aztec/src/hash.nr:
```diff
@@ -165,3 +165,16 @@ unconstrained fn compute_var_args_hash() {
     assert(hash == 0x19b0d74feb06ebde19edd85a28986c97063e84b3b351a8b666c7cac963ce655f);
 }
+
+#[test]
+unconstrained fn public_bytecode_commitment() {
+    let mut input = [0; MAX_PACKED_PUBLIC_BYTECODE_SIZE_IN_FIELDS];
+    let len = 99;
+    for i in 1..len + 1 {
+        input[i] = i as Field;
+    }
+    input[0] = (len as Field) * 31;
+    let hash = compute_public_bytecode_commitment(input);
+    dep::std::println(hash);
+    assert(hash == 0x16d621c3387156ef53754679e7b2c9be8f0bceeb44aa59a74991df3b0b42a0bf);
+}
```
I don't like the `println` or harcoded hash here, but wanted to keep consistent with the test above it (`compute_var_args_hash()`) and have an 'anchor' to help any future debugging

MirandaWood commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -84,7 +84,7 @@ namespace bc_hashing;
     // b) we are not at start, in which case we only read 2 fields, and => only increment pc by 31*2=62
     #[PC_INCREMENTS]
-    sel * (pc_index' - ((1 - LATCH_CONDITION) * (62 + pc_index + sel_not_start * 31))) = 0;
+    (sel + precomputed.first_row) * (pc_index' - ((1 - LATCH_CONDITION) * (62 + pc_index + sel_not_start * 31))) = 0;
```
This covers an edge case where the only place we did not constrain that `pc_index = 0` after a `latch` (<==> row is a start or inactive) was at row 0, since `sel` would be off. This was actually pretty bad as we could claim an incorrect hash for a real bytecode (see new test `NegativeSingleBytecodeHash`)

fcarreiro commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -15,15 +15,16 @@ include "../poseidon2_hash.pil";
+// bc_hash_0 = H([sep, 0x111, 0x222, 0x333, 0x444, 0x555, 0x666, 0x777])
+// bc_hash_1 = H([sep, 0xaaa, 0xbbb, 0xccc])
+// p_f = packed_fields:
+//   pc_index |  pc_index_1 |  pc_index_2 | bytecode_id |  p_f_0  |  p_f_1  |  p_f_2  |  output_hash  | latch | START | padding |
```
I was surprised that `output_hash` was always the same. I was thinking: how are the different rows of the hash being connected? could they be reordered? etc.

Then when looking at the poseidon2_hash pil I realized that it is exactly the propagation of that hash to all rows (in that gadget) + "infeasible to crack" which makes it secure. Maybe it would be good to at least remind (closer to the beginning of the file) that this is a quirk or the poseidon2_hash gadget (/remind how it works).

MirandaWood replied:
I was too! I'll add a comment, thanks :)

fcarreiro commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -66,68 +67,168 @@ namespace bc_hashing;
-    sel' * (start' - LATCH_CONDITION) = 0;
+    sel' * ((1 - sel_not_start') - LATCH_CONDITION) = 0;
 
     // Used as part of the lookup into bytecode decomposition
```
Would it simplify what follows if you could generate the hash of a bytecode as
```
poseidon2([SEPARATOR, 0, 0, field0, field1, field2...]);
```
?

If it simplifies things enough (not sure) it might be possible to consider changing it in the deployer contract?

MirandaWood replied:
It may, but I was hesistant to change the definition of the hash (at least in this PR) since it would require a lot of changes in the noir contract, protocol contract fixtures, ts, and may brick a load of tests. My logic was that a classic Miranda-sized PR probably wasn't worth simplifying 15 or so lines of PIL code 😅 
However I'd be happy to do a follow up - this would also make it easier to see the impact of 'lengthening' the hash by two fields?

fcarreiro commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -66,68 +67,168 @@ namespace bc_hashing;
+    #[START_IS_SEPARATOR]
+    START * (packed_fields_0 - constants.GENERATOR_INDEX__PUBLIC_BYTECODE) = 0;
+
+    #[GET_PACKED_FIELD_0]
```
I think it's ok to have 3 lookups, at least until the world proves otherwise. `bc_decomposition` is BIG, but `sel_packed` is 31 times smaller. I have some faith.

fcarreiro commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -66,68 +67,168 @@ namespace bc_hashing;
-    (1 - LATCH_CONDITION) * (incremental_hash' - output_hash) = 0;
+    // TODO(MW): add padding or number of fields/rounds to lookup?
+    #[POSEIDON2_HASH]
+    sel { latch, packed_fields_0, packed_fields_1, packed_fields_2, output_hash }
```
I was surprised that the length is not here, but the gadget lists it as an input (that goes into the IV). Seems unconstrained now.

MirandaWood replied:
Ah yep - I think I added a comment/TODO about this, we no longer use the length and I wasn't sure whether it was 100% safe to remove as an input (see `compute_public_bytecode_commitment`). I think that since we constrain that we hash all fields of the bytecode given by `bc_decomposition` we are covered?

IlyasRidhuan replied:
yeah this does feel like it's missing some things. i think we do need to add the input_len here so we guarantee the correct IV is being set, additionally you might need to even add in a round_count, that lines up with `poseidon2.num_perm_rounds_rem`. Otherwise a malicious prover might be able to pick and choose the the rounds to use.

Also do we need `output_hash` to be propagated down?

This is semi-related to this issue https://github.com/AztecProtocol/aztec-packages/issues/12738

MirandaWood replied:
Makes sense! Ty for linking the issue, I'd kind of assumed it was impossible for rounds to be tampered with (inserting extra ones, reordering etc) on the poseidon side so didn't initially add a lookup for any round related values (on the bc hashing side, since I increment pc, I know I can't skip/add chunks of values). I'll add the `input_len` here for the IV (and think about whether that means we can defer some of the ugly padding calcs to poseidon).

> Also do we need output_hash to be propagated down?

I don't think we _need_ to since my intention was that any lookup into bc hashing would use `latch` as the anchor (since it's only at `latch` we know we have reached the end of bytecode and hashed all fields), so we only 'care' about that final value of `output_hash`. Related to [another comment ](https://github.com/AztecProtocol/aztec-packages/pull/16756#discussion_r2329540573) below. Maybe the best approach is add documentation of this intention and keep the column as is?

IlyasRidhuan replied:
That does make sense but im a little concerned about essentially underconstraining the non-final value of `output_hash`

We need to guarantee that the invocation of this lookup on the latched row is indeed using the correct previous poseidon2 state (and so on). I'm concerned that a malicious prover might be able to do the following:

In bc_hashing, we have some  inputs that are `[1, 2, 3, 4, 5 6]` with output hash `x`. It just so happens that there is also a hash used in calldata hashing that is `[a, b, c, 4, 5, 6]` with output hash `y`. 

Can the attacker then use this lookup in the following way?
```
row 1: latch = 0,  inputs = [1, 2, 3], output_hash = x
row 2: latch = 1,  inputs = [4, 5, 6], output_hash = y
```

It would look like we hashed `[1, 2, 3, 4, 5, 6]` but we used the result of `[a, b, c, 4, 5, 6]`?

MirandaWood replied:
Done! One thought I had while writing this out - is it possible for a malicious prover to hash extra fields at the beginning? I don't think so but would like a sanity check if possible!

We constrain that all bytecode fields from the decomp trace are included (in order) in the bc_hashing trace. Each field is looked up into the poseidon trace. With the addition of `rounds_rem` we now know those fields are in the same order (i.e. haven't swapped some round rows). Also, with the addition of `input_len` and constraining it against the final bytes remaining stuff, we know that the IV is correct and (I think?) there are no extra fields in the poseidon preimage.

Though there's nothing constraining that the bc_hashing trace's start row corresponds to the poseidon_2 trace's start row, I think that since we know `input_len` is correct vs the bytecode itself, and `rounds_rem` is correct (= 1 at `latch`), _and_ these two values are linked at poseidon_2's start row, then no extra fields can be added at the beginning?

MirandaWood replied:
> It would look like we hashed [1, 2, 3, 4, 5, 6] but we used the result of [a, b, c, 4, 5, 6]?

Good point! I didn't consider this -  I'm going to test this now, if it's possible I'll add a condition as for `bytecode_id` to the circuit.

EDIT: yep - good catch, added a hash consistency relation + test, thanks!

IlyasRidhuan replied:
> Also, with the addition of input_len and constraining it against the final bytes remaining stuff, we know that the IV is correct and (I think?) there are no extra fields in the poseidon preimage.

you might want to check how `input_len` works in poseidon2 hash. It might only be constrained on the start row (indirectly thorough IV), it could be sensible to propagate.  Otherwise you might need to add something that ensures that there exists a row in poseidon2hash that is `{ start, input_len}`, having input_len alone probably isnt enough because it's unconstrained

IlyasRidhuan replied:
The reason this isnt really a problem for other invocations of poseidon2Hash in our code is because we are able to directly use `start` as the dest selector for hashes with 1 permutation round
```
sel { .., len } in poseidon2.start { ..., input_len }
```

MirandaWood replied:
> t might only be constrained on the start row (indirectly thorough IV), it could be sensible to propagate.

So yep, it does only appear to be constrained on the first row. The reason I wasn't concerned here is that in both traces (bc_hashing and poseidon), we constrain that `rounds_rem` decrements down to 1, and IIUC in the poseidon trace we must have a `start` row which links `input_len` to `rounds_rem`. Our bc hashing `input_len` is constrained at `latch` to be correct, so this should link everything together?

I see that if a `start` row does not exist in the poseidon trace with our correct `input_len`, that's a problem. But I think it must exist due to the latch/start checks, and even though its not propagated down, `num_perm_rounds_rem` is decremented down and looked up. I guess a possible issue is not propagating the same `input_len` per row in the bc hashing trace (like the output hash check I just added)?

fcarreiro commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -15,15 +15,16 @@ include "../poseidon2_hash.pil";
 // Finally, we copy through a lookup/permutation based on the selector these field elements
```
<!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'><i>[Re: line +5]</i></p><!--__GRAPHITE_HTML_TAG_END__-->
nit: we need the constants file (it's being included transitively now)
<!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16756?utm_source=unchanged-line-comment">Graphite</a>.</p><!--__GRAPHITE_HTML_TAG_END__-->

fcarreiro commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
Question: could there be multiple non-continuous bytecode_id entries? e.g. can you hash the same bytecode twice? in principle nothing wrong unless...

Could you have different output_hashes if you maybe choose to stop early etc? given that this trace is looked up into (I think) we should at least think about what could be the cases so that we know we are not giving the lookup too much choice.

MirandaWood replied:
Good q! I was thinking about this - ideally we would want to prove that at some point we do reach the end of the bytecode and use the `output_hash` given there, since it's only there (at `latch`) that we fully constrain everything vs the decomposition. But obviously there's no way to say in the pil "at some point further down there will be a `latch`". We do know that a `start` must follow a `latch`/row 0 and have `pc_index = 0`, then following pc indices must count up accordingly, so I think the beginning is covered.
Maybe the best thing is to make clear that it's always expected to lookup into this trace by `latch = 1`?

EDIT: Just to note that I added a comment about looking up by `latch`, plus we constrain more vs the poseidon lookup, so are a bit more robust here

fcarreiro reviewed (commented):
Please only consider my review as additional to @IlyasRidhuan since I didn't go into the fine details :)

IlyasRidhuan reviewed (approved):
Looks good! Nice job with the approach to handling padding. The hash seems a bit underconstrained so might need a closer look there

MirandaWood committed:
feat(avm!): Bytecode Hashing

## Constraining Bytecode Hashing

### Overview

Following elsewhere in the repo, we constrain that the bytecode commitment is the poseidon2 hash of the bytecode (encoded as fields) prepended by the special separator (`GENERATOR_INDEX__PUBLIC_BYTECODE`):
```
hash = poseidon2([GENERATOR_INDEX__PUBLIC_BYTECODE, FF(bytecode[0...31]), FF(bytecode[31...62]), ... ])
```

See `encode_bytecode` and `compute_public_bytecode_commitment` for C++ logic (unchanged by this PR).

### PIL

The general idea is that each row of `bc_hashing` corresponds to a row of `poseidon2`. It is the 'link' between `bc_decomposition` and `poseidon2`, proving that the hash given by `output_hash` is indeed a valid hash of the exact bytecode from `bc_decomposition` (with no extra or missing info), prepended by `GENERATOR_INDEX__PUBLIC_BYTECODE`.

The decomp circuit already helpfully sets the packed field every 31 bytes (looking forward), where `sel_packed = 1` and `packed_field` is the field given by packing the bytes from `pc` to `pc + 31` (starting at 0). Since each poseidon permutation takes in a chunk of 3 fields, each row of `bc_hashing` handles 3 packed fields. For now, I'm looking up each separately (for `i = 0, 1, 2`):
```
    #[GET_PACKED_FIELD_i]
    sel { pc_index_i, bytecode_id, packed_fields_i }
    in
    bc_decomposition.sel_packed { bc_decomposition.pc, bc_decomposition.id, bc_decomposition.packed_field };
```
where (generally) each `pc_index_i` is simply `pc_index_{i - 1} + 31`. Then each field is looked up into the hasher:

```
    #[POSEIDON2_HASH]
    sel { latch, packed_fields_0, packed_fields_1, packed_fields_2, output_hash }
    in poseidon2_hash.sel { poseidon2_hash.end, poseidon2_hash.input_0, poseidon2_hash.input_1, poseidon2_hash.input_2, poseidon2_hash.output };
```

Easy! Oh wait. Edge cases.

#### Bytecode Start

At the beginning of a new bytecode, we have `START = 1` and `pc_index = 0`. We must constrain that the first field hashed is the prepended separator:

```
    #[START_IS_SEPARATOR]
    START * (packed_fields_0 - constants.GENERATOR_INDEX__PUBLIC_BYTECODE) = 0;
```
 However, this means that `packed_fields_0` does not actually exist as a packed field in `bc_decomposition`, so the lookup `GET_PACKED_FIELD_0` will fail. To rectify this, at `START = 1`:
 - We have `sel_not_start` = 0 and skip the `GET_PACKED_FIELD_0` lookup
 - We set `pc_index_1 = 0`, so `GET_PACKED_FIELD_1` reads the first encoded field of the bytecode from `bc_decomposition` (and => `pc_index_2 = 31` and reads the second)
 - Instead of constraining that `pc_index' = pc_index + 93` (usual row), we only increment by 62
 - The poseidon lookup remains the same

#### Bytecode End

As before, once we read the end of some bytecode, we set `latch = 1` and constrain that the next row is a `start` (if exists). We have two main things to consider here:
 - Constrain that when `latch = 1`, we are at the end of decomposed bytes given by `bc_decomposition` (and read all of them)
 - Deal with padding in `poseidon2`
Both are dealt with in the `PADDING` section, since both require asserting we have 'trailing zeros' and have correctly read/set the packed fields.

Now that each row handles 3 fields, we may have 0, 1, or 2 padding fields which will not exist in `bc_decomposition`. To avoid the lookups bricking, I added two selectors:
 - `sel_not_padding_1` <==> `packed_field_1` is not a padding value => switch on the `GET_PACKED_FIELD_1` lookup
 - `sel_not_padding_2` <==> `packed_field_2` is not a padding value => switch on the `GET_PACKED_FIELD_2` lookup

(Note that `packed_field_0` can never be a padding value and must always be processed, and all padding selectors are gated by `latch`)

This now adds an attack vector where we could incorrectly claim real values are padding values in order to skip the lookup and => skip hashing bytecode fields. To avoid this I added `pc_at_final_field` which is the value of `pc` at the beginning of the last field e.g. if `packed_fields_1` is the last bytecode field, then `pc_at_final_field = pc_index_1`. This is constrained by the relation `PADDING_CORRECTNESS`.

From here, the simplest way to prove everything we need for padding/end of bytecode is to check that the `bc_decomposition.bytes_remaining` at `pc_at_final_field` is `< 31`. But this needs an evil range check. I don't want an evil range check on top of the 3 packed field lookups. Hence the following madness.

The decomposition trace processes windows of (currently) 37 bytes, and forces these bytes to be zero if they overshoot the end of the bytecode. Iff we have less than 37 bytes remaining to process, `bc_decomposition.sel_windows_gt_remaining = 1` => we can easily check that `bytes_remaining < 37` without a range check by just looking up this selector.

What about the remaining 6 bytes, you ask? These are all stored in `bc_decomposition.bytes_pc_plus_i`, so the final lookup becomes this monstrosity:
```
   #[CHECK_FINAL_BYTES_REMAINING]
    latch {
        pc_at_final_field,
        bytecode_id,
        sel /* =1 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */,
        precomputed.zero /* =0 */
    } in bc_decomposition.sel_packed {
        bc_decomposition.pc,
        bc_decomposition.id,
        bc_decomposition.sel_windows_gt_remaining,
        bc_decomposition.bytes_pc_plus_31,
        bc_decomposition.bytes_pc_plus_32,
        bc_decomposition.bytes_pc_plus_33,
        bc_decomposition.bytes_pc_plus_34,
        bc_decomposition.bytes_pc_plus_35,
        bc_decomposition.bytes_pc_plus_36
    };
```

But hey, we know we're at the final field at `pc_at_final_field` and there are no bytes remaining without a range check! Note that if the bytecode ends in valid zeros, I don't think we need to differentiate between these and 'padding' non-existent zeros because the final poseidon hash result would be the same.

To cover the poseidon considerations, we just need to make sure that the padding fields are indeed zero (`PADDED_BY_ZERO_i`) and if we claim that `packed_field_1` is a padding value, then so is `packed_field_2` (i.e. we can't have `[last_field, padding, some_crap]` at the end of a bytecode, `PADDING_CONSISTENCY`). If we have any padding at all, then `PADDING_2 = 1 - sel_not_padding_2` must be on, so I use this as the general padding selector when required.

#### Bytecode Start & End

Oh yes. These things can both be true. Isn't that fun! You might notice some ugly complexities in relations and this is probably why.

A much nicer way of showing `bytes_remaining < 31` without a range check would be to set some `pc_index_to_lookup = pc_at_final_field - 6`, then show that at `pc_index_to_lookup` that `bc_decomposition.sel_windows_gt_remaining = 1`. However, if both `latch` and `start` are true here then we have `pc_index = 0` and this would underflow.

I also made sure to use the `pc_index_i` columns to work out `pc_at_final_field` because they take into consideration the shift we do at `start` for the separator. Otherwise we could use something like `31*padding_amount` which would be a lot prettier (alas).

### TODOs/Qs

 - Should we explore using shifts to remove the 3 lookups into `bc_decomposition` for the packed fields?
 - Since the padding/latch conditions (I think?) constrain that we hash all bytes, the `bytecode_length` event property is unused, remove?
 - Is it worth converting the padding columns to a numerical `padding_amount` (=0, 1, or 2) and looking this up into `poseidon_2`?
